Add chainerx.elu

aksub99 opened the pull request:
Add `chainerx.elu`.

aksub99 commented:
@niboshi What could be the reason for the failing backward tests?

emcastillo commented on chainerx_cc/chainerx/routines/CMakeLists.txt:
```diff
@@ -1,5 +1,9 @@
+=======
+    activation.cc
+>>>>>>> Add chainerx.elu
     binary.cc
```
There is a syntax error in cmake here

emcastillo replied:
The build is failing due to an incorrect CMakeLists,
Please fix it :)

aksub99 replied:
Thanks @emcastillo !

aksub99 commented:
@niboshi Could you please review?

niboshi commented:
@aksub99 @emcastillo Thank you for the PR & comments!

niboshi commented on chainerx_cc/chainerx/routines/activation.cc:
```diff
@@ -0,0 +1,21 @@
+
```
Extra blank line?

niboshi commented on chainerx_cc/chainerx/routines/activation.cc:
```diff
@@ -0,0 +1,21 @@
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/explog.h"
+#include "chainerx/routines/indexing.h"
+#include "chainerx/scalar.h"
```
* Move ` #include "chainerx/routines/activation.h" ` to the first line of the file.
* It seems some includes are not necessary (constant.h, dtype.h, kernels/math.h, <vector>)

niboshi commented on chainerx_cc/chainerx/routines/activation.cc:
```diff
@@ -0,0 +1,21 @@
+namespace chainerx {
+Array Elu(const Array& x, Scalar alpha) {
+    Array zero = ZerosLike(x, x.device());
+    return Where(x > zero, x, alpha * Expm1(x));
```
`x > zero` could be replaced with `x > 0` if `operator>` supports scalars.
Could you write a TODO?

niboshi commented on chainerx_cc/chainerx/routines/activation.h:
```diff
@@ -0,0 +1,14 @@
+#include <nonstd/optional.hpp>
+
+#include "chainerx/array.h"
+#include "chainerx/constant.h"
```
* Some includes are unnecessary. (cstdint, optional, constant)
* Add scalar.h.

niboshi commented on tests/chainerx_tests/unit_tests/routines_tests/test_activation.py:
```diff
@@ -75,6 +75,37 @@ def forward_xp(self, inputs, xp):
+))
+class TestElu(UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    alpha = 2.0
```
Please parameterize `alpha`, including a negative value, zero and "unspecified".

aksub99 commented:
@niboshi Sorry for the delay in addressing your comments. I have made all the suggested fixes. The tests are failing because of mismatch in `dtypes` of outputs and gradients. Could the gradients be having a different `dtype` because of `alpha`? If so, how can I change the type of  `alpha` to be the same as the output `dtype`(Since `alpha` might be a tuple)?

niboshi commented on tests/chainerx_tests/unit_tests/routines_tests/test_activation.py:
```diff
@@ -72,6 +73,44 @@ def forward_xp(self, inputs, xp):
+            negzero_indices = y <= 0
+            y[negzero_indices] = self.alpha * numpy.expm1(y[negzero_indices])
+            return y
+        return xp.elu(a, self.alpha)
```
Thank you for the fix.
Please also test "unspecified" alpha.
You could implement similarly to existing tests like [test_connection.py](https://github.com/chainer/chainer/blob/v7.0.0a1/tests/chainerx_tests/unit_tests/routines_tests/test_connection.py#L427) (define `Unspecified` class and parameterize it as `alpha_range`).

aksub99 replied:
Sure. Will modify the parametrizations accordingly.

niboshi commented:
Please check Travis failure, too.

aksub99 commented:
@niboshi I have parametrized `alpha` with Unspecified.
I'm unsure of how to fix the travis failures. I thought the mismatch in dtypes might be because of alpha's dtype but I guess I was mistaken.

kshitij12345 commented on tests/chainerx_tests/unit_tests/routines_tests/test_activation.py:
```diff
@@ -72,6 +78,49 @@ def forward_xp(self, inputs, xp):
+        if isinstance(self.alpha_range, tuple):
+            l, u = self.alpha_range
+            self.alpha = random.uniform(l, u)
+        elif self.alpha_range in (None, Unspecified):
```
Is `Unspecified` necessary?
With 
`'alpha_range': [(-2.0, 0.0), (0.0, 2.0),  0.0, 1.0]`

We will have
```Python
if isinstance(self.alpha_range, tuple):
  l, u = self.alpha_range
  self.alpha = random.uniform(l, u)
else:
  self.alpha = self.alpha_range
```
Which will have the same behavior as it does now.

aksub99 commented:
@niboshi I have fixed the travis failures and rebased. Please review:)

niboshi commented:
Thanks!

niboshi commented on tests/chainerx_tests/unit_tests/routines_tests/test_activation.py:
```diff
@@ -144,6 +152,60 @@ def func(self, xp, a):
+        if isinstance(self.alpha_range, tuple):
+            l, u = self.alpha_range
+            self.alpha = random.uniform(l, u)
+        elif self.alpha_range in (None, Unspecified):
```
Can you remove `None`?

niboshi commented on tests/chainerx_tests/unit_tests/routines_tests/test_activation.py:
```diff
@@ -144,6 +152,60 @@ def func(self, xp, a):
+            negzero_indices = y <= 0
+            y[negzero_indices] = self.alpha * numpy.expm1(y[negzero_indices])
+            return y
+        elif self.alpha_range in (None, Unspecified):
```
ditto.

niboshi commented on chainerx_cc/chainerx/python/routines.cc:
```diff
@@ -625,12 +625,13 @@ void InitChainerxManipulation(pybind11::module& m) {
 void InitChainerxActivation(pybind11::module& m) {
     m.def("clipped_relu", [](const ArrayBodyPtr& x, Scalar z) { return MoveArrayBody(ClippedRelu(Array{x}, z)); }, "x"_a, "z"_a = 20.0);
     m.def("crelu", [](const ArrayBodyPtr& x, int8_t axis) { return MoveArrayBody(CRelu(Array{x}, axis)); }, "x"_a, "axis"_a = 1);
+    m.def("elu", [](const ArrayBodyPtr& x, float_t alpha) { return MoveArrayBody(Elu(Array{x}, alpha)); }, "x"_a, "alpha"_a = 1.0);
```
`float_t` -> `double`?

niboshi commented on chainerx_cc/chainerx/routines/activation.cc:
```diff
@@ -38,6 +39,14 @@ Array CRelu(const Array& x, int8_t axis) {
     return Relu(concat);
 }
 
+Array Elu(const Array& x, float_t alpha) {
```
ditto.

niboshi commented on chainerx_cc/chainerx/routines/activation.h:
```diff
@@ -9,6 +9,8 @@ Array ClippedRelu(const Array& x, Scalar z);
 
 Array CRelu(const Array& x, int8_t axis);
 
+Array Elu(const Array& x, float_t alpha = 1.0);
```
ditto.

aksub99 commented:
@niboshi I have made the required changes.

niboshi commented:
Thanks.
Jenkins, test this please

aksub99 commented:
The test failures seem to be caused by `group_normalization` which is unrelated to this PR. Am I correct?

niboshi commented:
@aksub99 
Yes, you're right.
Thank you for the PR!
